# Exercises

# Do your work for this exercise in a jupyter notebook or python script named tidy_data. Save this work in your classification-exercises repo. Then add, commit, and push your changes.

# 1. Attendance Data

# Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

# You should end up with something like this:


# name
# Billy    0.5250
# Jane     0.6875
# John     0.9125
# Sally    0.7625
# Name: grade, dtype: float64

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

In [2]:
attendance = pd.read_csv('untidy-data/attendance.csv')
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [3]:
attendance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  4 non-null      object
 1   2018-01-01  4 non-null      object
 2   2018-01-02  4 non-null      object
 3   2018-01-03  4 non-null      object
 4   2018-01-04  4 non-null      object
 5   2018-01-05  4 non-null      object
 6   2018-01-06  4 non-null      object
 7   2018-01-07  4 non-null      object
 8   2018-01-08  4 non-null      object
dtypes: object(9)
memory usage: 416.0+ bytes


In [4]:
#rename columns
attendance.columns
attendance.columns = ['name', '2018-01-01', '2018-01-02', '2018-01-03', '2018-01-04', '2018-01-05', '2018-01-06', '2018-01-07', '2018-01-08']
attendance


name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [5]:
# Melting data except name
attendance= attendance.melt(id_vars = 'name', var_name = 'date', value_name = 'roll_call')
attendance

name        date roll_call
0   Sally  2018-01-01         P
1    Jane  2018-01-01         A
2   Billy  2018-01-01         A
3    John  2018-01-01         P
4   Sally  2018-01-02         T
5    Jane  2018-01-02         P
6   Billy  2018-01-02         T
7    John  2018-01-02         T
8   Sally  2018-01-03         T
9    Jane  2018-01-03         T
10  Billy  2018-01-03         A
11   John  2018-01-03         H
12  Sally  2018-01-04         H
13   Jane  2018-01-04         T
14  Billy  2018-01-04         A
15   John  2018-01-04         P
16  Sally  2018-01-05         P
17   Jane  2018-01-05         T
18  Billy  2018-01-05         H
19   John  2018-01-05         P
20  Sally  2018-01-06         A
21   Jane  2018-01-06         T
22  Billy  2018-01-06         T
23   John  2018-01-06         T
24  Sally  2018-01-07         T
25   Jane  2018-01-07         A
26  Billy  2018-01-07         P
27   John  2018-01-07         P
28  Sally  2018-01-08         T
29   Jane  2018-01-08         T
30  Billy  2018-01-08         T
31   John  2018-01-08         P

In [6]:
def calc_point(roll_call):
    if roll_call == "P":
        point = 1
    elif roll_call == "H":
        point = 0.5
    elif roll_call == "T":
        point = 0.9
    else:
        point = 0
        
    return point
        

In [7]:
attendance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   name       32 non-null     object
 1   date       32 non-null     object
 2   roll_call  32 non-null     object
dtypes: object(3)
memory usage: 896.0+ bytes


In [8]:
attendance['attendance_point'] = attendance.roll_call.apply(calc_point)
attendance

name        date roll_call  attendance_point
0   Sally  2018-01-01         P               1.0
1    Jane  2018-01-01         A               0.0
2   Billy  2018-01-01         A               0.0
3    John  2018-01-01         P               1.0
4   Sally  2018-01-02         T               0.9
5    Jane  2018-01-02         P               1.0
6   Billy  2018-01-02         T               0.9
7    John  2018-01-02         T               0.9
8   Sally  2018-01-03         T               0.9
9    Jane  2018-01-03         T               0.9
10  Billy  2018-01-03         A               0.0
11   John  2018-01-03         H               0.5
12  Sally  2018-01-04         H               0.5
13   Jane  2018-01-04         T               0.9
14  Billy  2018-01-04         A               0.0
15   John  2018-01-04         P               1.0
16  Sally  2018-01-05         P               1.0
17   Jane  2018-01-05         T               0.9
18  Billy  2018-01-05         H               0.5
19   John  2018-01-05         P               1.0
20  Sally  2018-01-06         A               0.0
21   Jane  2018-01-06         T               0.9
22  Billy  2018-01-06         T               0.9
23   John  2018-01-06         T               0.9
24  Sally  2018-01-07         T               0.9
25   Jane  2018-01-07         A               0.0
26  Billy  2018-01-07         P               1.0
27   John  2018-01-07         P               1.0
28  Sally  2018-01-08         T               0.9
29   Jane  2018-01-08         T               0.9
30  Billy  2018-01-08         T               0.9
31   John  2018-01-08         P               1.0

In [9]:
attendance.groupby(['name']).mean()

attendance_point
name                   
Billy            0.5250
Jane             0.6875
John             0.9125
Sally            0.7625

# ALTERNATIVE METHOD:

In [10]:
attendance['point'] = attendance.roll_call.replace({'P': 1, 'A': 0, 'H': 0.5, 'T': 0.9})
attendance


name        date roll_call  attendance_point  point
0   Sally  2018-01-01         P               1.0    1.0
1    Jane  2018-01-01         A               0.0    0.0
2   Billy  2018-01-01         A               0.0    0.0
3    John  2018-01-01         P               1.0    1.0
4   Sally  2018-01-02         T               0.9    0.9
5    Jane  2018-01-02         P               1.0    1.0
6   Billy  2018-01-02         T               0.9    0.9
7    John  2018-01-02         T               0.9    0.9
8   Sally  2018-01-03         T               0.9    0.9
9    Jane  2018-01-03         T               0.9    0.9
10  Billy  2018-01-03         A               0.0    0.0
11   John  2018-01-03         H               0.5    0.5
12  Sally  2018-01-04         H               0.5    0.5
13   Jane  2018-01-04         T               0.9    0.9
14  Billy  2018-01-04         A               0.0    0.0
15   John  2018-01-04         P               1.0    1.0
16  Sally  2018-01-05         P               1.0    1.0
17   Jane  2018-01-05         T               0.9    0.9
18  Billy  2018-01-05         H               0.5    0.5
19   John  2018-01-05         P               1.0    1.0
20  Sally  2018-01-06         A               0.0    0.0
21   Jane  2018-01-06         T               0.9    0.9
22  Billy  2018-01-06         T               0.9    0.9
23   John  2018-01-06         T               0.9    0.9
24  Sally  2018-01-07         T               0.9    0.9
25   Jane  2018-01-07         A               0.0    0.0
26  Billy  2018-01-07         P               1.0    1.0
27   John  2018-01-07         P               1.0    1.0
28  Sally  2018-01-08         T               0.9    0.9
29   Jane  2018-01-08         T               0.9    0.9
30  Billy  2018-01-08         T               0.9    0.9
31   John  2018-01-08         P               1.0    1.0

In [11]:
attendance.groupby(['name']).mean()

attendance_point   point
name                           
Billy            0.5250  0.5250
Jane             0.6875  0.6875
John             0.9125  0.9125
Sally            0.7625  0.7625

# 2. Coffee Levels

- a. Read the coffee_levels.csv file.

In [12]:
coffee_levels = pd.read_csv('untidy-data/coffee_levels.csv')
coffee_levels.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [13]:
coffee_levels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   hour           30 non-null     int64  
 1   coffee_carafe  30 non-null     object 
 2   coffee_amount  30 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 848.0+ bytes


- b. Transform the data so that each carafe is in it's own column.

In [14]:
new_data = coffee_levels.pivot_table(index = ['hour'],
                                       columns = 'coffee_carafe',
                                       values = 'coffee_amount').reset_index()
new_data

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

- c. Is this the best shape for the data?

In [15]:
new_data.T

0         1          2          3          4          5  \
coffee_carafe                                                                   
hour           8.000000  9.000000  10.000000  11.000000  12.000000  13.000000   
x              0.816164  0.451018   0.843279   0.335533   0.898291   0.310711   
y              0.189297  0.521502   0.023163   0.235529   0.017009   0.997464   
z              0.999264  0.915990   0.144928   0.311495   0.771947   0.398520   

                       6          7          8          9  
coffee_carafe                                              
hour           14.000000  15.000000  16.000000  17.000000  
x               0.507288   0.215043   0.183891   0.391560  
y               0.058361   0.144644   0.544676   0.594126  
z               0.864464   0.436364   0.280621   0.436677

# 3. Cake Recipes

- a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [16]:
cake_recipes = pd.read_csv('untidy-data/cake_recipes.csv')
cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [17]:
cake_recipes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   recipe:position  8 non-null      object 
 1   225              8 non-null      float64
 2   250              8 non-null      float64
 3   275              8 non-null      float64
 4   300              8 non-null      float64
dtypes: float64(4), object(1)
memory usage: 448.0+ bytes


- b. Tidy the data as necessary.

In [18]:
cake_recipes[['recipe', 'position']] =cake_recipes['recipe:position'].str.split(':', expand = True)
cake_recipes.head()

recipe:position        225        250        275        300 recipe position
0        a:bottom  61.738655  53.912627  74.414730  98.786784      a   bottom
1           a:top  51.709751  52.009735  68.576858  50.228470      a      top
2        b:bottom  57.095320  61.904369  61.196980  99.248541      b   bottom
3           b:top  82.455004  95.224151  98.594881  58.169349      b      top
4        c:bottom  96.470207  52.001358  92.893227  65.473084      c   bottom

In [19]:
cake_recipes = cake_recipes.drop(columns = 'recipe:position')
cake_recipes

225        250        275        300 recipe position
0  61.738655  53.912627  74.414730  98.786784      a   bottom
1  51.709751  52.009735  68.576858  50.228470      a      top
2  57.095320  61.904369  61.196980  99.248541      b   bottom
3  82.455004  95.224151  98.594881  58.169349      b      top
4  96.470207  52.001358  92.893227  65.473084      c   bottom
5  71.306308  82.795477  92.098049  53.960273      c      top
6  52.799753  58.670419  51.747686  56.183110      d   bottom
7  96.873178  76.101363  59.571620  50.971626      d      top

In [21]:
cake_recipes_melt = cake_recipes.melt(id_vars = ['recipe', 'position'], var_name = 'temperature', value_name = 'score')
cake_recipes_melt.head(6)

recipe position temperature      score
0      a   bottom         225  61.738655
1      a      top         225  51.709751
2      b   bottom         225  57.095320
3      b      top         225  82.455004
4      c   bottom         225  96.470207
5      c      top         225  71.306308

- c. Which recipe, on average, is the best? recipe b

In [23]:
cake_recipes_melt.groupby(['recipe']).score.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: score, dtype: float64

- d. Which oven temperature, on average, produces the best results? 275

In [24]:
cake_recipes_melt.groupby('temperature').score.mean()

temperature
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: score, dtype: float64

- e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [26]:
cake_recipes_melt.groupby(['recipe', 'position', 'temperature']).score.mean()

recipe  position  temperature
a       bottom    225            61.738655
                  250            53.912627
                  275            74.414730
                  300            98.786784
        top       225            51.709751
                  250            52.009735
                  275            68.576858
                  300            50.228470
b       bottom    225            57.095320
                  250            61.904369
                  275            61.196980
                  300            99.248541
        top       225            82.455004
                  250            95.224151
                  275            98.594881
                  300            58.169349
c       bottom    225            96.470207
                  250            52.001358
                  275            92.893227
                  300            65.473084
        top       225            71.306308
                  250            82.795477
                  275   